src/store/message-store-level.ts

```diff
@@ -77,19 +82,150 @@ export class MessageStoreLevel implements MessageStore {
+    filters: Filter[],
+    messageSort?: MessageSort,
+    pagination?: Pagination,
+    options?: MessageStoreOptions
+  ): Promise<{ messages: GenericMessage[], paginationMessageCid?: string }> {
     options?.signal?.throwIfAborted();
 
     const messages: GenericMessage[] = [];
+    const resultIds = await this.index.query(filters.map(f => ({ ...f, tenant })), options);
 
-    const resultIds = await this.index.query({ ...filter, tenant }, options);
-
+    // as an optimization for large data sets, we are finding the message object which matches the paginationMessageCid here.
+    // we can use this within the pagination function after sorting to determine the starting point of the array in a more efficient way.
+    let paginationMessage: GenericMessage | undefined;
     for (const id of resultIds) {
       const message = await this.get(tenant, id, options);
       if (message) { messages.push(message); }
+      if (pagination?.messageCid && pagination.messageCid === id) {
+        paginationMessage = message;
+      }
+    }
+
+    if (pagination?.messageCid !== undefined && paginationMessage === undefined) {
+      return { messages: [] }; //if paginationMessage is not found, do not return any results
     }
 
-    return messages;
+    const sortedRecords = await MessageStoreLevel.sortMessages(messages, messageSort);
+    return this.paginateRecords(sortedRecords, paginationMessage, pagination);
+  }
+
+  private async paginateRecords(
+    messages: GenericMessage[],
+    paginationMessage?: GenericMessage,
+    pagination: Pagination = { }
+  ): Promise<{ messages: GenericMessage[], paginationMessageCid?: string } > {
+    const { limit } = pagination;
+    if (paginationMessage === undefined && limit === undefined) {
+      return { messages }; // return all without pagination pointer.
+    }
+
+    // this is an optimization hack, we are passing the pagination message object for an easier lookup
+    // since we know this object exists within the the array if passed, we can assume that it will always have a value greater than -1
+    // if no pagination message is passed, we start from the beginning of the array, so we set the "cursor" to -1
+    const cursorIndex = paginationMessage ? messages.indexOf(paginationMessage) : -1;
+
+    // the first element of the returned results is always the message immediately following the cursor.
+    const start = cursorIndex > 0 ? cursorIndex + 1 : 0;
+    const end = limit === undefined ? undefined : limit + start;
+    const results = messages.slice(start, end);
+
+    // we only return a pagination MessageCid cursor if there are more results
+    const hasMoreResults = end !== undefined && end < messages.length;
+    let paginationMessageCid: string|undefined;
+    if (hasMoreResults) {
+      // we extract the cid of the last message in the result set.
+      const lastMessage = results.at(-1);
+      paginationMessageCid = lastMessage ? await Message.getCid(lastMessage) : undefined;
```

Comment:
Sanity: why would the `lastMessage` be `undefined` if there are more results?

```suggestion
      paginationMessageCid = await Message.getCid(lastMessage!);
```

Reply:
Push minor changes for this, see if you like/hate it. Please resolve if it looks good to you!